File: interval_sum/solutions/all_in_memory.cpp
#include<cassert>

const int MAXN = 100000;
int N;
int val[MAXN];
int **memo_risp;

void get_sequence(int n, int *s) {
   N = n;
   for (int i=0; i<N; i++) {
     val[i] = s[i];
   }
   memo_risp = new int*[N];
   for(int i=0; i<N; i++)
     memo_risp[i] = new int[N];
   for (int i=0; i<N; i++) {
       memo_risp[i][i] = s[i];
     for (int j=i+1; j<N; j++) {
       memo_risp[i][j] = memo_risp[i][j-1]+s[j];
     }
   }  
}

int interval_sum(int a, int b) {
  assert( a >= 0 );
  assert( b >= a -1 );
  assert( b < N );
  if(b<a)
    return 0;
  return memo_risp[a][b];
}

File: interval_sum/solutions/all_in_memory.py
val = []
memo_risp = []

def get_sequence(n, s):
    global N
    N = n
    for i in range(N):
        val.append(s[i])
        memo_risp.append([s[i]])
        for j in range(i+1,N):
            memo_risp[-1].append(memo_risp[-1][-1]+s[j])

def interval_sum(a, b):
    assert a >= 0
    assert b >= a -1
    assert b < N
    if b < a:
        return 0
    return memo_risp[a][b-a]
